File: backend/utils/validators.py
"""
Data validation utility functions.
Following the common-modules.md specification.
"""
import re
from typing import List


def validate_year(year: int) -> bool:
    """
    Validate year is within acceptable range.

    Args:
        year: Year to validate

    Returns:
        True if valid, False otherwise
    """
    return 2000 <= year <= 2100


def validate_email(email: str) -> bool:
    """
    Validate email format.

    Args:
        email: Email string to validate

    Returns:
        True if valid, False otherwise
    """
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_file_extension(filename: str, allowed_extensions: List[str]) -> bool:
    """
    Validate file extension is in allowed list.

    Args:
        filename: Name of file to validate
        allowed_extensions: List of allowed extensions (e.g., ['xlsx', 'xls'])

    Returns:
        True if valid, False otherwise
    """
    if not filename or '.' not in filename:
        return False

    extension = filename.rsplit('.', 1)[-1].lower()
    return extension in [ext.lower() for ext in allowed_extensions]


def validate_file_size(file_size: int, max_size_mb: int) -> bool:
    """
    Validate file size is within limit.

    Args:
        file_size: File size in bytes
        max_size_mb: Maximum allowed size in MB

    Returns:
        True if valid, False otherwise
    """
    max_size_bytes = max_size_mb * 1024 * 1024
    return file_size <= max_size_bytes

File: frontend/src/utils/validators.js
/**
 * Client-side validation utility functions.
 * Following the common-modules.md specification.
 */

/**
 * Validate email format.
 *
 * @param {string} email - Email string to validate
 * @returns {boolean} True if valid email format
 */
export const isValidEmail = (email) => {
  if (!email) return false;
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
};

/**
 * Validate file extension.
 *
 * @param {File} file - File object to validate
 * @param {Array<string>} allowedExtensions - Array of allowed extensions (e.g., ['xlsx', 'xls'])
 * @returns {boolean} True if file extension is allowed
 */
export const isValidFileExtension = (file, allowedExtensions) => {
  if (!file || !file.name) return false;

  const extension = file.name.split('.').pop().toLowerCase();
  return allowedExtensions.map(ext => ext.toLowerCase()).includes(extension);
};

/**
 * Validate file size.
 *
 * @param {File} file - File object to validate
 * @param {number} maxSizeMB - Maximum size in MB
 * @returns {boolean} True if file size is within limit
 */
export const isValidFileSize = (file, maxSizeMB) => {
  if (!file) return false;

  const maxSizeBytes = maxSizeMB * 1024 * 1024;
  return file.size <= maxSizeBytes;
};

/**
 * Validate required field.
 *
 * @param {any} value - Value to validate
 * @returns {boolean} True if value is not empty
 */
export const isRequired = (value) => {
  if (value === null || value === undefined) return false;
  if (typeof value === 'string') return value.trim().length > 0;
  return true;
};

/**
 * Validate number range.
 *
 * @param {number} value - Number to validate
 * @param {number} min - Minimum value
 * @param {number} max - Maximum value
 * @returns {boolean} True if number is within range
 */
export const isInRange = (value, min, max) => {
  const num = Number(value);
  if (isNaN(num)) return false;
  return num >= min && num <= max;
};

/**
 * Validate year.
 *
 * @param {number} year - Year to validate
 * @returns {boolean} True if year is valid (2000-2100)
 */
export const isValidYear = (year) => {
  return isInRange(year, 2000, 2100);
};

export default {
  isValidEmail,
  isValidFileExtension,
  isValidFileSize,
  isRequired,
  isInRange,
  isValidYear,
};
